l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol

```diff
@@ -127,13 +121,34 @@ contract L2NativeTokenVault is IL2NativeTokenVault, NativeTokenVault {
         }
     }
 
+    function _ensureTokenDeployedInnerLegacyToken(
+        bytes32 _assetId,
+        address _originToken,
+        bytes memory _erc20Data,
+        address _expectedToken,
+        address _l1LegacyToken
+    ) internal {
+        _assetIdCheck(L1_CHAIN_ID, _assetId, _originToken);
+
+        /// token is a legacy token, no need to deploy
+        if (_l1LegacyToken != _originToken) {
+            revert AddressMismatch(_originToken, _l1LegacyToken);
+        }
+
+        tokenAddress[_assetId] = _expectedToken;
+        assetId[_expectedToken] = _assetId;
+    }
+
     /// @notice Deploys the beacon proxy for the L2 token, while using ContractDeployer system contract.
     /// @dev This function uses raw call to ContractDeployer to make sure that exactly `l2TokenProxyBytecodeHash` is used
     /// for the code of the proxy.
     /// @param _salt The salt used for beacon proxy deployment of L2 bridged token.
     /// @return proxy The beacon proxy, i.e. L2 bridged token.
-    function _deployBeaconProxy(bytes32 _salt) internal virtual override returns (BeaconProxy proxy) {
-        if (address(L2_LEGACY_SHARED_BRIDGE) == address(0)) {
+    function _deployBeaconProxy(
+        bytes32 _salt,
+        uint256 _tokenOriginChainId
+    ) internal virtual override returns (BeaconProxy proxy) {
+        if (address(L2_LEGACY_SHARED_BRIDGE) == address(0) && _tokenOriginChainId != L1_CHAIN_ID) {
```

Comment:
this looks weird, if L2_LEGACY_SHARED_BRIDGE) == address(0) and _tokenOriginChainId == L1_CHAIN_ID you will go into the `else` statement and use `L2_LEGACY_SHARED_BRIDGE`

Reply:
true, its ||